### ISSUE TITLE
config: `self-hosted-runner.labels` is parsed but no rule reads it

### ISSUE THREAD

**Issue by on-keyday** [ultra-supara/sisakulint#610]
**TL;DR** — The `self-hosted-runner.labels` key is accepted, parsed, printed in the `-debug`
configuration line, and echoed back when the configuration file is regenerated. No rule reads it.
Of the four keys in `Config`, it is the only one with no rule consumer. Setting it changes nothing
— including for the job shape it would most plausibly exist for: a job that reaches a self-hosted
runner through a custom label alone.

The question this raises is what the key is for, so the issue ends with two shapes rather than one
recommended patch.

## Findings

All source references are at `624588a2` (`main`, 2026-08-16).

### 1. Three of the four configuration keys reach a rule; `labels` does not

Every rule receives the configuration:

```go
func (rule *BaseRule) UpdateConfig(config *Config) {
```

— [rule.go#L74](https://github.com/sisaku-security/sisakulint/blob/624588a28aa8c10ab529fd0caaac00243d14e2cf/pkg/core/rule.go#L74)

Three keys are read through that field, inside the rule that owns them:

- `action-list` — `for _, pattern := range rule.userConfig.actionListRegex {`
  ([actionlist.go#L34](https://github.com/sisaku-security/sisakulint/blob/624588a28aa8c10ab529fd0caaac00243d14e2cf/pkg/core/actionlist.go#L34))
- `config-variables` — `v = rule.userConfig.ConfigVariables`
  ([exprchecker.go#L412](https://github.com/sisaku-security/sisakulint/blob/624588a28aa8c10ab529fd0caaac00243d14e2cf/pkg/core/exprchecker.go#L412))
- `secret-exfiltration.allowed-hosts` — `for _, h := range rule.userConfig.SecretExfiltration.AllowedHosts {`
  ([secretexfiltration.go#L286](https://github.com/sisaku-security/sisakulint/blob/624588a28aa8c10ab529fd0caaac00243d14e2cf/pkg/core/secretexfiltration.go#L286))

`SelfHostedRunner.Labels` has two readers, neither of them a rule:

```go
parts = append(parts, fmt.Sprintf("self-hosted-runner.labels: %v", c.SelfHostedRunner.Labels))
```

— [config.go#L47](https://github.com/sisaku-security/sisakulint/blob/624588a28aa8c10ab529fd0caaac00243d14e2cf/pkg/core/config.go#L47), inside `Config.String()`, which surfaces only in the `-debug` line

```go
if existingConfig != nil && len(existingConfig.SelfHostedRunner.Labels) > 0 {
```

— [actionlist_generator.go#L138](https://github.com/sisaku-security/sisakulint/blob/624588a28aa8c10ab529fd0caaac00243d14e2cf/pkg/core/actionlist_generator.go#L138), which writes the value back out when the configuration file is regenerated

### 2. The rule matches the literal label only

```go
if label != nil && strings.EqualFold(label.Value, "self-hosted") {
```

— [selfhostedrunnersrule.go#L71](https://github.com/sisaku-security/sisakulint/blob/624588a28aa8c10ab529fd0caaac00243d14e2cf/pkg/core/selfhostedrunnersrule.go#L71); the matrix path compares the same way
([#L108](https://github.com/sisaku-security/sisakulint/blob/624588a28aa8c10ab529fd0caaac00243d14e2cf/pkg/core/selfhostedrunnersrule.go#L108))

The file contains no reference to `userConfig`. A runner group is reported on its own, but a label
is reported only when it is spelled `self-hosted`.

### 3. Measured: the value reaches the debug line and stops there

Two projects, identical configuration files apart from `labels`, tool built at `624588a2`:

| | `labels: []` | `labels: [self-hosted, linux-large]` |
|---|---:|---:|
| `runs-on: self-hosted` | 1 × `self-hosted-runner` | 1 × `self-hosted-runner` |
| `runs-on: [linux-large]` | 0 | **0** |
| `action-list` (control, same file) | 1 | 1 |

The control matters: the configuration file is being read — the `action-list` entry in it fires in
both projects — so the zeros are not a configuration that failed to load. With `-debug` the value
is visibly present:

```
setting configuration: Config{self-hosted-runner.labels: [self-hosted linux-large], action-list: [actions/checkout@*]}
```

So a user who declares `linux-large` as their self-hosted label, and whose workflow says
`runs-on: [linux-large]`, gets no finding — while the declaration sits in the configuration the
tool just printed.

### 4. The key predates the rule

- The key arrived with configuration file support: `generate init config`, 2023-10-25
  ([ba3a557](https://github.com/sisaku-security/sisakulint/commit/ba3a557))
- The rule arrived over two years later: `feat: add self-hosted-runner rule (#276)`, 2026-01-18
  ([0954d1b](https://github.com/sisaku-security/sisakulint/commit/0954d1b)) — and did not wire the
  configuration in

Until #578 the generated configuration described the key as letting sisakulint "verify that these
labels are correctly configured". That sentence is gone; the key is still there.

**Recommended change:** decide what the key is for. Two shapes, and this report does not pick
between them:

- **Wire it.** Treat a label listed in `self-hosted-runner.labels` as self-hosted for detection.
  The plumbing already exists — the rule can read `userConfig` the way three other rules do — so
  this is a condition added to the label comparison. It closes the custom-label case in Finding 3.
- **Drop the key.** `parseConfig` uses `yaml.Unmarshal` without `KnownFields`, so an unknown key is
  ignored: removing the field does not break configuration files that already carry it.

## Scope & limitations

- How often workflows reach self-hosted runners by a custom label alone was not measured, so the
  size of the gap in Finding 3 is unknown. What is shown is that the declared form is not reported.
- Wiring the key makes this rule's output depend on the configuration file: the same workflow would
  lint differently with and without it. `action-list` already behaves that way, but
  `self-hosted-runner` currently fires with no configuration at all, so the change is more visible
  here.
- A declared label is self-report. A user who lists a label that actually resolves to a
  GitHub-hosted larger runner would get a false positive under the wiring option.
- Only the four keys of `Config` were traced. Whether other declared-but-unread settings exist
  elsewhere was not checked.

## Reproduction

Tool pinned at `624588a28aa8c10ab529fd0caaac00243d14e2cf` (`main`, 2026-08-16):

```bash
git clone https://github.com/sisaku-security/sisakulint && cd sisakulint
git checkout 624588a28aa8c10ab529fd0caaac00243d14e2cf && go build -o sisakulint ./cmd/sisakulint
SL=$PWD/sisakulint

# sisakulint walks up for an ancestor holding both .git and .github/workflows
setup() {   # setup <dir> <runs-on value>
  rm -rf "$1"; mkdir -p "$1/.github/workflows"; git -C "$1" init -q
  cat > "$1/.github/workflows/ci.yaml" <<YAML
name: CI
on: [push]
jobs:
  build:
    runs-on: $2
    timeout-minutes: 10
    permissions:
      contents: read
    steps:
      - uses: unauthorized/suspicious-action@v1
YAML
}

config() {  # config <dir> <labels block>
  cat > "$1/.github/sisakulint.yaml" <<YAML
self-hosted-runner:
  labels: $2

config-variables: null

action-list:
  - actions/checkout@*

secret-exfiltration:
  allowed-hosts: []
YAML
}

count() { (cd "$1" && "$SL" 2>&1) | grep -c "\[$2\]"; }

setup /tmp/lit self-hosted
setup /tmp/cus '[linux-large]'

config /tmp/lit '[]';                          count /tmp/lit self-hosted-runner   # 1
config /tmp/lit '[self-hosted, linux-large]';  count /tmp/lit self-hosted-runner   # 1
config /tmp/cus '[self-hosted, linux-large]';  count /tmp/cus self-hosted-runner   # 0
count /tmp/cus action-list                                                          # 1  <- config is live
(cd /tmp/lit && "$SL" -debug 2>&1) | grep -o 'setting configuration: Config{[^}]*}'
```

## Status

Draft recommendation from an internal review; not accepted internally yet. Findings 1, 2 and 4 are
quotes from the pinned source and commit metadata; Finding 3 is from the reproduction above.

🤖 Generated with [Claude Code](https://claude.com/claude-code)
